src/main/scala/com/microsoft/hyperspace/actions/DeleteOnReadAction.scala

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (2020) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.actions
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.SparkSession
+
+import com.microsoft.hyperspace.HyperspaceException
+import com.microsoft.hyperspace.index._
+import com.microsoft.hyperspace.telemetry.{AppInfo, DeleteOnReadActionEvent, HyperspaceEvent}
+
+/**
+ * Refresh index by updating list of excluded source data files and index signature
+ * in index metadata.
+ * Note this Refresh Action only fixes an index metadata w.r.t deleted source data files
+ * and does not consider new source data files (if any).
+ * If some original source data file(s) are removed between previous version of index and
+ * now, this Action refreshes index as follows:
+ * 1. Deleted source data files are identified.
+ * 2. New index fingerprint is computed w.r.t latest source data files.
+ * 3. IndexLogEntry is updated by modifying list of excluded source data files and
+ * index fingerprint, computed in above steps.
+ *
+ * @param spark SparkSession.
+ * @param logManager Index LogManager for index being refreshed.
+ * @param dataManager Index DataManager for index being refreshed.
+ */
+class DeleteOnReadAction(
+    spark: SparkSession,
+    logManager: IndexLogManager,
+    dataManager: IndexDataManager)
+    extends RefreshDeleteActionBase(spark, logManager, dataManager)
+    with Logging {
+
+  final override protected def event(appInfo: AppInfo, message: String): HyperspaceEvent = {
+    DeleteOnReadActionEvent(appInfo, logEntry.asInstanceOf[IndexLogEntry], message)
+  }
+
+  override def validate(): Unit = {
+    super.validate()
+    if (deletedFiles.toSet.equals(previousIndexLogEntry.excludedFiles.toSet)) {
```

Comment:
You can create `private lazy val newExludedFiles: Seq[String] = deletedFiles diff previousIndexLogEntry.excludedFiles` and use it here (`if (newExludedFiles.isEmpty)`) and in line 103?

Reply:
Sure, done.